Refactor user-defined channel match

MrPetovan opened the pull request:
Follow-up to https://github.com/friendica/friendica/pull/13806

- ~~Move throwaway full-text search feature to the Database class~~
- Move disposable full-text search feature to a Database-namespaced class
- ~~Invert the boolean logic in favor of a positive match in `Repository\UserDefinedChannel->getMatchingChannelUsers`~~
- Convert generic `foreach` into an `array_filter` call in `Repository\UserDefinedChannel->getMatchingChannelUsers`
- Fix return value of `Repository\UserDefinedChannel->match`

MrPetovan committed:
Add throwaway full-text search feature in a dedicated Database class

- Add explicit return type to UserDefinedChannels->current() to help IDE auto-completion

MrPetovan committed:
Refactor user-defined channel match

- Remove disposable full-text search feature from UserDefinedChannels repository
- Fix return value of Repository\UserDefinedChannel->match

annando commented on src/Content/Conversation/Repository/UserDefinedChannel.php:
```diff
@@ -221,62 +219,30 @@ public function getMatchingChannelUsers(string $searchtext, string $language, ar
+			$uids = [];
+
+			foreach ($this->select(['uid' => array_column($users, 'uid'), 'publish' => true, 'valid' => true]) as $channel) {
+				if (
```
That `if`construct is ... massive. Fun fact: in my company, this would be rejected during the code review because this is hardly to debug.

MrPetovan replied:
Let me see if I can make it look smaller.

MrPetovan replied:
I understand what you're saying but I believe the previous code was hard to read because of the systematic negative conditions (and the duplicated conditions on `$uids`).

What we want to do is pick the User Id if all the conditions match which is what I wrote. It's indeed harder to debug individual conditions in my version but I'm not sure the "return early" philosophy is welcome in a loop.

annando commented on src/Database/Database.php:
```diff
@@ -310,6 +310,57 @@ public function databaseName(): string
+	 * @return bool
+	 * @throws \Exception
+	 */
+	public function performThrowawayFullTextSearch(string $haystack, string $needle): bool
```
I still don't really like the name. For me the name should describe the outcome, but not the way it performs the outcome. So this "throwaway" is not need in the function name, I think.

MrPetovan replied:
I see you point, but I feel like we should still indicate why we have a separate dedicated set of methods for fulltext search when it can just be added to a regular SQL condition. Maybe "throwaway" isn't the correct term, but I believe we still need a term to separate this feature from the regular way of performing a full-text search in a SQL condition.

annando replied:
It could always happen that at some point in the future we would be able to do this whole stuff in memory, without the need for this table. Then we had to rename the function, although for the user nothing would change.

MrPetovan replied:
It's true, hence my question about the naming, if not throwaway, then what? "OneOff"?

MrPetovan replied:
"disposable" ?

annando commented on src/Database/Database.php:
```diff
@@ -310,6 +310,57 @@ public function databaseName(): string
+	 * @return mixed
+	 * @throws \Exception
+	 */
+	public function prepareThrowawayFullTextSearch(string $haystack, \Closure $callback)
```
Wouldn't it be better to make a class out of it? Then the entry could be created in the constructor and be removed in the deconstructor.

MrPetovan replied:
It's as valid, especially since we are using a dedicated table for this. The implementation details would look very similar.

annando replied:
At least for me, it would be easier to use (and understand).

MrPetovan replied:
I can look into that.

MrPetovan commented:
Renamed "throwaway" to "disposable", moved the feature to a dedicated class, rewrote the matching condition inside an `array_filter` callback which allowed me to keep the condition separated without losing meaning.

Of course, this is all untested, but it looked good to my IDE's static analyzer.

annando commented on src/Database/DisposableFullTextSearch.php:
```diff
@@ -0,0 +1,59 @@
+		$this->db = $database;
+
+		// Maximum value is indicated by the INT UNSIGNED type of the check-full-text-search.pid field
+		$this->identifier = random_int(0, pow(2, 32) - 1);
```
Why don't you use the pid?

MrPetovan replied:
if you create two DisposableFullTextSearch object in the same script execution, the first object destruction would delete both searchtext rows, before the second object destruction is called, leading to unexpected behavior. If each object has it's own random identifier ensures the expected behavior of the specific searchtext deleted only when the corresponding object is destroyed.

annando replied:
Since this class is used by us and it isn't some third party class, we can document the behaviour. The problem with a random number is that it is random and that it can happen that the same number can be returned at the same time.

MrPetovan replied:
Yes, there's always a chance of a collision, but it is cosmically low given the range of the `identifier` field.

I tried to find a way to generate unique integer identifiers using both `getmypid()` and `spl_object_id()` but the entropy is abysmal and it would more likely lead to collisions.

annando replied:
The chance of a collision is even lower when we just ensure that we never initiate two instances of the class.

annando replied:
To handle possible collisions, you could check the return value of the insert command and then could do the random stuff.

MrPetovan replied:
The return value will be `true` because there's no auto increment `id` field (nor should there be).

> The chance of a collision is even lower when we just ensure that we never initiate two instances of the class.

There's no need to make it into a Singleton class. The current randomness is enough to prevent collisions at the scale this class is meant to be used. Of course, if the `identifier` was converted into a string, it would be possible to aggregate the server hostname, the script PID and the object ID in a way that would be unique for each instance of this class even on a node with multiple frontend servers and a single database backend.

annando replied:
When you remove the `Database::INSERT_UPDATE` in the insert command, then it will fail on a duplicate entry. And then the return value of the `Insert` command will be `false`. So then you could try again with a different value. Or you could add an additional primary key that contains the timestamp returned by `microtime`, since you cannot have got the same timestamp with the same pid.

MrPetovan replied:
On a distributed system (multiple frontends with a single database backend), you can definitely have the same pid and the same timestamp.

And the timestamp returned by microtime would either need to be a string, or a large integer if multiplied by 10000. In either case there are more reliable ways to either generate entropy, or ensure uniqueness.

annando replied:
We currently don't support distributed systems. When we do so, we will have to add an additional factor also in the worker queue, like some "host id".

This class is called from two places. And it is unlikely that it will be called at much more places. Also it is just some internally used class. This means that we can be sure that there won't be multiple instances at the same time.

MrPetovan replied:
I implemented a loop with the insert because I don't want to change the DB schema for this.

annando commented:
Is it possible for you to test it? Otherwise it has to wait at least until Sunday, where I most likely will have some time.

MrPetovan commented:
I wouldn't know where to start, I've never used user-defined channels before. I'm fine waiting.

annando commented:
I could guide you through the user defined channels.

MrPetovan commented:
Thanks, but you might as well test on your end where you already have the framework for it.

annando commented:
I will be on holiday for the next two weeks, so I will have got time for that. Can you solve the conflict?

MrPetovan commented:
Done.

annando commented:
@MrPetovan Do you know a quicker way to fetch the branch without the need to add another remote?

MrPetovan commented:
No, but I can help you with the latter:
```
git remote add MrPetovan https://github.com/MrPetovan/friendica
git fetch MrPetovan
git checkout MrPetovan/task/refactor-throwaway-fulltext-search
```

annando commented:
Okay, then I will do so tomorrow. The last days I fought with my local internet setup. BTW: When you wanna feel like some BOFH and want to use a device that doesn't forgive even the slightest errors, you should use internet router from MikroTik :grimacing:

annando commented on src/Content/Conversation/Repository/UserDefinedChannel.php:
```diff
@@ -221,62 +219,53 @@ public function getMatchingChannelUsers(string $searchtext, string $language, ar
+		$disposableFullTextSearch = new DisposableFullTextSearch($this->db, $searchtext);
 
-		$uids = [];
+		$filteredChannels = $this->select(['uid' => array_column($users, 'uid'), 'publish' => true, 'valid' => true])->filter(
```
What is this `filter` function and how exactly does it work?

MrPetovan replied:
`Collection->filter()` works the same way as `array_filter()`: it returns a filtered collection depending on the return value of a callback function that is performed on each entity in the collection. If the return value is truthy, the entity is kept, if the value is falsy, the entity is discarded from the returned collection.

annando commented:
I'm currently testing it on my machine. Please have got a look at my two questions.

MrPetovan commented:
Thanks, I couldn't see the second question though. 👀

MrPetovan committed:
Add documentation to DisposableFullTextSearch->idenfier field

MrPetovan committed:
Ensure identifier uniqueness in Disposable FullTextSearch

annando commented on src/Database/DisposableFullTextSearch.php:
```diff
@@ -0,0 +1,62 @@
+			// would delete both check-full-text-search rows before the second object destruction is called, leading to unexpected behavior.
+			// Maximum value is indicated by the INT UNSIGNED type of the check-full-text-search.pid field
+			$this->identifier = random_int(0, pow(2, 32) - 1);
+		} while($this->db->insert('check-full-text-search', ['pid' => $this->identifier, 'searchtext' => $haystack]));
```
Can you add some limit, so that there isn't an endless loop when the database becomes unavailable? I guess that for example `10` should be a good number, since it is extremely unlikely that you would had 10 hits. Otherwise of course you could have a quick check via `exists`.

MrPetovan replied:
`exists()` will also return `false` if the database isn't available, so there's no benefit to use a second query.

annando replied:
When `exists` returned `false`, then the system would perform some `insert` that would fail as well. But this wouldn't lead to some endless loop like it would now. Especially when this was called via a worker, then in the worst case that loop would continue for several hours.

MrPetovan replied:
Got it. What about the behavior when we can't insert the search text row? Are we fine with silently failing any subsequent match calls or should we throw an exception?

annando replied:
I'm fine with silently failing, since the only condition where this should happen is some database issue. And such a situation will cause a lot of problems at many places.

MrPetovan committed:
Improve uniqueness loop by adding an exists() call in DisposableFullTextSearch

annando commented on src/Database/DisposableFullTextSearch.php:
```diff
@@ -0,0 +1,67 @@
+		do {
+			// Maximum value is indicated by the INT UNSIGNED type of the check-full-text-search.pid field
+			$this->identifier = random_int(0, pow(2, 32) - 1);
+		} while($this->db->exists('check-full-text-search', ['pid' => $this->identifier, 'searchtext' => $haystack]));
```
Please remove `, 'searchtext' => $haystack` here. Also please add a space after `while` to comply with the code standards.

MrPetovan replied:
Done.

MrPetovan committed:
Fix exists() condition in DisposableFullTextSearch
